contracts/core/EntryPoint.sol

```diff
@@ -141,6 +145,27 @@ contract EntryPoint is IEntryPoint, StakeManager {
         _compensate(beneficiary, collected);
     }
 
+    /**
+     * perform estimation of enture UserOperation.
+     * this method will always revert. it performs full validation of the UserOperation,
+     * but passing NO_VALIDATION_AGGREGATOR to the account.validateUserOperation, to signal it
+     * to ignore signature failure.
+     * This way, this method can be used to estimate entire execution without requiring the user
+     * to sign the request.
+     * Note that in order to collect the the success/failure of the target call, it must be executed
+     * with trace enabled to track the emitted events.
+     */
+    function estimateExecution(UserOperation calldata op) external {
+
+        UserOpInfo memory opInfo;
+
+        _validatePrepayment(0, op, opInfo, NO_VALIDATION_AGGREGATOR);
```

Comment:
We execute _validatePrepayment+_executeUserOp but not the rest of the overhead of handleOps.  Therefore it can be used for simulation but not for gas estimation (unless the fixed overhead is added).

Reply:
It does almost all the same on-chain gas calculations as handleOps (there is a slight difference, since there is no loop over an array)
It does differ in preVerificationGas:
- it doesn't account for bundle size, can modify greatly preVerificationGas.
- it doesn't account for aggregator gas (again, it can't be calculated, and must be added to preVerificationGas)